### PR TITLE
fix(observability): Emit wrapped metrics for reused components

### LINKED ISSUE #4454: Unify span context when components are nested
According to https://github.com/timberio/vector/pull/4181#issuecomment-700391110, the metrics labels from nested components will not reflect the parent component.

For example, the `humio_logs` sink wraps the `splunk_hec` sink. The `splunk_hec` sink [defines a `SplunkEventSent`](https://github.com/timberio/vector/blob/3507ed2d00d95011e0f9a19b24da02a9014caa54/src/internal_events/splunk_hec.rs#L14) which emits the `events_processed` metric. This metric will include labels that identify the `splunk_hec` sink instead of the `humio_logs` sink. To a user this is confusing, since they defined a `humio_logs` sink, not a `splunk_hec` sink.

### PR DESCRIPTION
**Pull request by ktff**
Closes #4454 

Replaces previous approach in #5174 which had a flaw. Most relevant comment https://github.com/timberio/vector/pull/5174#issuecomment-738353342. 

Approach of this PR is to wrap reused components with a layer that sets thread local flag on entry, and resets it on exit. Then during emitting we check for this flag and call either `emit_metrics` or `emit_metrics_wrapped` accordingly. With this to reuse a component it should be wrapped with `wrap_emit()`, implementing `trait WrapEmit` if needed.

Shared code can be dealt with in the same way, just a special care of what is running in and outside of wrapped code is needed.

### Tried alternatives
Instead of setting the flag and then checking for it, entering/exiting additional span was tried out but the performance cost was quite severe. @MOZGIII 

### Todo

- [x] Do a more thorough audit of where we are reusing components.

- [x] Should we implement `emit_metrics_wrapped` for all problematic metrics now, or delegate it to when a component gets reused? (EDIT: implemented now since they are just empty functions)

- [ ] Follow up on https://github.com/timberio/vector/pull/5424#discussion_r537414234

### CODE REVIEW

File: src/sinks/humio/metrics.rs [hunk 1]
```diff
@@ -69,7 +70,7 @@ impl GenerateConfig for HumioMetricsConfig {
 #[typetag::serde(name = "humio_metrics")]
 impl SinkConfig for HumioMetricsConfig {
     async fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
-        let mut transform = self.transform.clone().build().await?;
+        let mut transform = self.transform.clone().build().await?.wrap_emit();
```
Comment (binarylogic) [inline src/sinks/humio/metrics.rs:73]: I have a feeling we'll forget to do this in the future, especially as we start to expand our integrations. Is there a way to force the developer to explicitly choose? Such as a required argument in `build()` or something along those lines.

Please disregard if this doesn't make any sense 😄

Comment (ktff) [reply src/sinks/humio/metrics.rs:73]: Good idea, but we can do even better. I think it would be possible to completely automate this, but I'll need to explore this idea a bit.

Comment (ktff) [reply src/sinks/humio/metrics.rs:73]: We could detect build inside other build:
   * Will not catch builds outside of build method, but this shouldn't happen since errors of sub builds should be propagated through master build so this should lead contributors to build inside other build. (But this is a lot of shoulds) 
   * This can be done by wrapping build future to set thread local flag from which we can tell if this is main build or sub build.
   * Requires a layer of traits for 
         * `SourceConfig`
         * `TransformConfig`
         * `SinkConfig`
     in which we would modify built future to track the flag and wrap built component with `wrap_emit` or not depending on the flag.

cc. @lukesteensen

Comment (ktff) [reply src/sinks/humio/metrics.rs:73]: Scratch that, once problematic cases like `new_relic_logs` are taken into consideration automatic detection will run into the same problems as #5174 .

That said, I think finding a way to force the developer to explicitly choose will be better served in a separate issue which can then build upon this PR.

Comment (MOZGIII) [reply src/sinks/humio/metrics.rs:73]: > That said, I think finding a way to force the developer to explicitly choose will be better served in a separate issue which can then build upon this PR.

I agree, this will allow us to progress in the meantime, and unblock other works that depend on this.

**Review by lukesteensen — COMMENTED** [review]
I could have sworn that we already pulled up the shared event emission for transforms up to the topology level (/cc @Hoverbear). Even if we haven't yet, is there a reason we can't do that? It should be relatively straightforward to do [here](https://github.com/timberio/vector/blob/ed31012bdf8fcf11b4862b49eaf28c55d306f02a/src/topology/builder.rs#L112-L133) if I'm not missing something.

If we solve the issue for transforms, it seems like we should be able to do something simpler for just sources and sinks, if there's still a problem at all. The most common case there is not nesting, where both would emit events, but one as a facade over the other, where only the internal one would emit events. Would this remain a problem at that point?

(Apologies for any context I may be missing here)

**Comment by ktff** [comment]
> It should be relatively straightforward to do here

Yep. The idea has showed up several times now but I only saw it as a part of a solution.

> If we solve the issue for transforms, it seems like we should be able to do something simpler for just sources and sinks, if there's still a problem at all. 

True. For this PR I've made complete audit of what is being reused where and can confirm that currently only reused transforms are the issue, plus a few other cases that were dealt with by removing the extra emitted metric.

> The most common case there is not nesting, where both would emit events, but one as a facade over the other, where only the internal one would emit events.

True. There was only one case of real nesting, and that was in `kubernetes` source that reused `file` source, but that changed at some point as it now only reuses the file server.

> (Apologies for any context I may be missing here)

No no, thanks for making the comment, it seams I made a dragon out of a rabbit. 

---
To summarize, to solve current double emitted counts:
1. Raise `processed_events_total` for transforms to topology.
2. Fix few other cases.

That's it. When we are confronted with a good reason to nest components that is problematic, we can revisit this.

@lukesteensen @MOZGIII

File: src/internal_events/vector.rs [hunk 2]
```diff
@@ -2,18 +2,6 @@ use super::InternalEvent;
...
-    pub byte_size: usize,
-}
-
-impl InternalEvent for VectorEventSent {
-    fn emit_metrics(&self) {
-        counter!("processed_events_total", 1);
-        counter!("processed_bytes_total", self.byte_size as u64);
-    }
-}
-
```
Comment (MOZGIII) [inline src/internal_events/vector.rs:16]: I assume this is unused and thus removed?

File: src/internal_events/prometheus.rs [hunk 3]
```diff
@@ -129,19 +131,13 @@ impl InternalEvent for PrometheusRemoteWriteSnapError {
 
 #[derive(Debug)]
 pub struct PrometheusRemoteWriteReceived {
-    pub byte_size: usize,
```
Comment (MOZGIII) [inline src/internal_events/prometheus.rs:132]: Why is this removed?

File: src/internal_events/prometheus.rs [hunk 4]
```diff
@@ -129,19 +131,13 @@ impl InternalEvent for PrometheusRemoteWriteSnapError {
...
 
 impl InternalEvent for PrometheusRemoteWriteReceived {
     fn emit_logs(&self) {
         debug!(message = "Received remote_write events.", count = ?self.count);
     }
-
-    fn emit_metrics(&self) {
-        counter!("processed_events_total", self.count as u64);
-        counter!("processed_bytes_total", self.byte_size as u64);
-    }
```
Comment (MOZGIII) [inline src/internal_events/prometheus.rs:144]: Why aren't metrics sent here?

File: src/internal_events/wasm/event_processing.rs [hunk 5]
```diff
@@ -66,14 +66,21 @@ impl InternalEvent for EventProcessingProgress {
...
+
+    fn emit_metrics_wrapped(&self) {
         counter!("wasm_event_processing_total", 1,
             "component_role" => self.role.as_const_str(),
             "state" => self.state.as_const_str(),
         );
         match self.state {
-            State::Completed => counter!("processed_events_total", 1,
-                "component_role" => self.role.as_const_str(),
-            ),
```
Comment (MOZGIII) [inline src/internal_events/wasm/event_processing.rs:76]: I suggest we emit a component-specific counter here instead of removing this entirely

File: src/internal_events/wrap.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
+use crate::{
+    transforms::{FunctionTransform, TaskTransform, Transform},
+    Event,
+};
+use futures01::{Poll, Stream};
+use std::cell::Cell;
+
+thread_local!(static WRAPPED: Cell<bool> = Cell::new(false));
```
Comment (MOZGIII) [inline src/internal_events/wrap.rs:8]: This won't work with `emit`s occurring at `tokio::spawn` since it uses a thread pool. In general, with the architecture as we have currently, we very rarely can assume the code will execute on the same thread reliably - so I don't think this is a good idea.
We can try hooking this into the `tracing::span` as I did for passing those labels into the metrics. However, that has some performance cost.

**Comment by MOZGIII** [comment]
> I could have sworn that we already pulled up the shared event emission for transforms up to the topology level (/cc @Hoverbear).

I recall someone (possibly even it was you) tried this and reported there were an issues with this approach.

> There was only one case of real nesting, and that was in `kubernetes` source that reused `file` source, but that changed at some point as now it only reuses the file server.

True. It was using the file server since the early days of the original PR, but I have considered wrapping the file source very early on.

There is a case of nesting at the `kubernetes_logs` source to this day - the file server, the parser and then partial message merger all do emit, in addition to the emit in the source itself. This results in numerous `event_processed` bumps.

**Comment by lukesteensen** [comment]
@ktff 

> To summarize, to solve current double emitted counts we can:
> 
>   1. Raise `processed_events_total` for transforms to topology.
> 
>   2. Fix few other cases.
> 
> 
> That's it. When we are confronted with a good reason to nest components that is problematic, we can revisit this.

Sounds good to me! Thanks for working through this.

---

@MOZGIII 

> I recall someone (possibly even it was you) tried this and reported there were an issues with this approach.

I have a memory of working on it with @Hoverbear and it turning out nicely, which is why I thought it was already the case. But I must have made that up or maybe it got lost in a draft PR somewhere. Either way, I'm optimistic that there will not be an issue but we'll have to try to know for sure.

> There is a case of nesting at the `kubernetes_logs` source to this day - the file server, the parser and then partial message merger all do emit, in addition to the emit in the source itself. This results in numerous `event_processed` bumps for a single event.

This case is one source and two transforms, correct? So removing the transform events will help. If we're still emitting once from the file source and once from the k8s source, we can customize the event emitter that is passed down to the file server. It doesn't emit directly since it's a different crate.

File: src/internal_events/vector.rs (hunk 2 above)
Comment (ktff) [reply src/internal_events/vector.rs:16]: Yep. More accurately `vector` sink was emitting this but the underlying `BytesSink` also emits those metrics so I removed the one in `vector` sink.

Comment (ktff) [reply src/internal_events/vector.rs:16]: `SocketEventsSent` is emitted in `BytesSink`.

File: src/internal_events/prometheus.rs (hunk 3 above)
Comment (ktff) [reply src/internal_events/prometheus.rs:132]: Because the `emit_metrics` that was using it was removed.

File: src/internal_events/prometheus.rs (hunk 4 above)
Comment (ktff) [reply src/internal_events/prometheus.rs:144]: `prometheus_remote_write` source is using shared `HttpSource` which is emitting those metrics with `HTTPEventsReceived` so this `emit_metrics` was deleted to remove double counting.

**Comment by ktff** [comment]
Closing in favor of #5492.
